File: NumbersDisappeared.py
# Time Complexity - O(N)

# Space Complexity - O(1)

# Did this code successfully run on Leetcode : No

# Any problem you faced while coding this : No

# Approach
# I use stqte change wherein I multiply the values by -1 whose index is abs(value) - 1.
# The answer is the indices of positive elements + 1

class Solution:
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        if nums == None or len(nums) == 0:
            return []

        result = []
        for i in range(len(nums)):
            index = abs(nums[i]) - 1
            if index < len(nums) and nums[index] > 0:
                nums[index] *= -1

        for i in range(len(nums)):
            if nums[i] > 0:
                result.append(i + 1)

        return result

File: NumbersDisappeared.cpp
// Time Complexity : o(n)
// Space Complexity : o(n)
// Did this code successfully run on Leetcode : YES
// Any problem you faced while coding this : NO


class Solution {
public:
    vector<int> findDisappearedNumbers(vector<int>& nums) {
        vector<int> ans;
        int n = nums.size();
        for(int i=0;i<n;i++) {
            int gotoIdx = abs(nums[i])-1;
            if(nums[gotoIdx] > 0) {
                nums[gotoIdx] = nums[gotoIdx] * -1;
            }
        }
        for(int i=0;i<n;i++) {
            if(nums[i] > 0) {
                ans.push_back(i+1);
            }
        }
        return ans;
    }
};